modules/standard/IO.chpl

```diff
@@ -8095,39 +8097,93 @@ private proc readBytesOrString(ch: fileReader, ref out_var: ?t, len: int(64)) :
-                                    iostringstyleInternal.data_toeof:int(64),
-                                    ch._channel_internal, tx,
-                                    lenread, uselen);
-    } else {
-      var save_style: iostyleInternal = ch._styleInternal();
-      var style: iostyleInternal = ch._styleInternal();
-      style.string_format = QIO_STRING_FORMAT_TOEOF;
-      ch._set_styleInternal(style);
-
-      if t == string {
-        err = qio_channel_scan_string(false,
-                                      ch._channel_internal, tx,
-                                      lenread, uselen);
+    // compute a guess as to the size to read based on the file length.
+    // this is only a guess & it's possible it will be out of date
+    // by the time we actually read.
+    var guessReadSize = 0;
+    {
+      var fp: qio_file_ptr_t = nil;
+      qio_channel_get_file_ptr(ch._channel_internal, fp);
+      var fileLen:int(64) = -1;
+      if fp {
+        fileLen = qio_file_length_guess(fp);
       }
-      else {
-        err = qio_channel_scan_bytes(false,
-                                     ch._channel_internal, tx,
-                                     lenread, uselen);
+      if pos >= 0 && fileLen >= 1 {
+        guessReadSize = fileLen - pos;
+      }
+      // limit the size to read by uselen
+      if guessReadSize > uselen {
+        guessReadSize = uselen;
       }
-      ch._set_styleInternal(save_style);
     }
 
-    var tmp = t.createAdoptingBuffer(tx, length=lenread);
+    // Note: remainder of this function assumes that the file data
+    // is in the same string encoding as the result (UTF-8). If/when
+    // fileReader supports other encodings, this will need to be updated.
+
+    // proactively allocate 'guessReadSize'
+    var buff: bufferType = nil;
+    var buffSz = 0;
+    var n:c_ssize_t = 0; // how many bytes have we read into buff?
+    assert(guessReadSize >= 0);
+    (buff, buffSz) = bufferAlloc(guessReadSize+1); // room for trailing \0
+
+    // then try to read repeatedly until we have read 'uselen' or reach EOF
+    while n < uselen {
+      var locErr:errorCode = 0;
+      var amtRead:c_ssize_t = 0;
+      if n >= buffSz {
+        // if we need more room in the buffer, grow it
+        var requestSz = 2*buffSz;
+        if requestSz < 16 then requestSz = 16;
+        (buff, buffSz) = bufferEnsureSize(buff, buffSz, requestSz);
+      }
+      assert(n < buffSz);
+      locErr = qio_channel_read(false, ch._channel_internal,
+                                buff[n], // read starting with data here
+                                min(uselen - n, buffSz - n),
+                                amtRead);
+      n += amtRead;
+      if locErr {
+        // reached EOF or other error so we need to stop
+        err = locErr;
+        break;
+      }
+    }
+
+    // add the trailing \0
+    (buff, buffSz) = bufferEnsureSize(buff, buffSz, n+1);
+    buff[n] = 0;
```

Comment:
I suggest reducing the size in `qio_channel_read` by 1 so we do not need to increase it here. I am thinking to pass `buffSize-n-1` to `min()` above in ine 8158.

Reply:
It would be unusual for this line to actually do anything, but it is there to make it easier to check that the corner cases are correct.

We cannot address the problem by decreasing the size of the read because we need to actually read whatever data is there. So if we have to read M bytes, we need M+1 memory to store these and also the trailing null byte. But we can't simply decide to read M-1 bytes.

Space for the trailing null byte is normally included due to

```
(buff, buffSz) = bufferAlloc(guessReadSize+1); // room for trailing \0
```

If guessReadSize is 0 etc, then we will be doubling the buffer as we go, in which case it is very likely that 1 byte will be available after reading, but if not `(buff, buffSz) = bufferEnsureSize(buff, buffSz, n+1);` will add room for it.

Reply:
Off-issue, we agreed that the approach you are describing is possible. However, I couldn't get it working without introducing correctness problems, and I don't think it's worth the complexity.